### PR TITLE
F#60 new demo

### PR DESCRIPTION
**Pull request by Dejin1**

### CODE REVIEW

File: Assets/Scripts/Map/TriggerSystem/Trigger Objects/Portal/PortalBehaviour.cs [hunk 1]
```diff
@@ -12,14 +13,50 @@
...
+    public float portalWait = 5;
+    bool timerStart = false;
+    bool portalActivated = false;
 
     public void ActivatePortal()
     {
-        GameObject camera = GameObject.Find("Main Camera");
-            float portalX = counterPortal.transform.position.x;
-            float portalY = counterPortal.transform.position.y;
+        if (timerStart && Time.time > startTime + portalWait)
```
Comment (massieb) [inline Assets/Scripts/Map/TriggerSystem/Trigger Objects/Portal/PortalBehaviour.cs]: add opening and closing brackets

File: Assets/Scripts/Map/TriggerSystem/Trigger Objects/Portal/PortalBehaviour.cs [hunk 2]
```diff
@@ -12,14 +13,50 @@
 public class PortalBehaviour : MonoBehaviour {
     //private PlayerDetection playerDetection; // Removed to try Trigger System.
     public GameObject counterPortal;
+    public Animator portalAnimator;
+    float startTime = 0;
+    public static string Idle = "Idle";
+    public static string Closing = "Closing";
+    public float portalWait = 5;
```
Comment (massieb) [inline Assets/Scripts/Map/TriggerSystem/Trigger Objects/Portal/PortalBehaviour.cs]: what is this variable? 5 ticks, seconds, milliseconds, frames?
I think its a portalOpeningDelayMs?

File: Assets/Scripts/Map/TriggerSystem/Trigger Objects/Portal/PortalBehaviour.cs [hunk 3]
```diff
@@ -12,14 +13,50 @@
...
-            float portalY = counterPortal.transform.position.y;
+        if (timerStart && Time.time > startTime + portalWait)
+            portalActivated = true;
+    }
+    
+    void Update()
+    {
+        if (portalAnimator.GetBool(Idle))
+        {
+            if (timerStart == false)
```
Comment (massieb) [inline Assets/Scripts/Map/TriggerSystem/Trigger Objects/Portal/PortalBehaviour.cs]: if(!timerStart)

File: Assets/Scripts/Map/Weather/Weather.cs [hunk 4]
```diff
@@ -255,10 +256,10 @@ void DeActivateRain()
 
     void RainAction()
     {
-        foreach (Rigidbody2D rb in rbs)
+        foreach (Rigidbody2D rb in rbs.ToList())
         {
-            //remove all rbs that get destroyed
-            if (rb == null || rb.gameObject.activeInHierarchy == false)
+                //remove all rbs that get destroyed
```
Comment (massieb) [inline Assets/Scripts/Map/Weather/Weather.cs]: abbreviations can be written in full in comments. --> what are rbs

File: Assets/Scripts/Map/Weather/Weather.cs [hunk 5]
```diff
@@ -255,10 +256,10 @@ void DeActivateRain()
...
+                if (rb == null || rb.gameObject.activeInHierarchy == false)
```
Comment (massieb) [inline Assets/Scripts/Map/Weather/Weather.cs]: add opening and closing brackets

File: Assets/Scripts/Player/JinPlayerMovement.cs [hunk 6]
```diff
@@ -230,6 +230,7 @@ public void OnLanding()
     void FixedUpdate()
     {
        previousHeight = currentHeight;
+       if (canMove)
```
Comment (massieb) [inline Assets/Scripts/Player/JinPlayerMovement.cs]: add opening and closing brackets

File: Assets/Audio/AudioScripts/AudioFadeIn.cs [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    public static IEnumerator FadeIn(AudioSource audioSource, float FadeTime)
+    {
+        float startVolume = 0.1f;
+        float maxVolume = audioSource.volume;
+        audioSource.volume = startVolume;
+
+        while (audioSource != null && startVolume < maxVolume)
+        {
+            audioSource.volume += audioSource.volume * Time.deltaTime / FadeTime;
+            yield return null;
```
[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Audio/AudioScripts/AudioFadeIn.cs]: No I can do waitforseconds I just like it xD but ill change it

File: Assets/Audio/AudioScripts/AudioFadeIn.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+        }
+        if (audioSource != null)
```
[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Audio/AudioScripts/AudioFadeIn.cs]: Yeah good catch will do

File: Assets/Scripts/Map/TriggerSystem/Trigger Objects/Portal/PortalBehaviour.cs (hunk 2 above)
Comment (Dejin1) [reply Assets/Scripts/Map/TriggerSystem/Trigger Objects/Portal/PortalBehaviour.cs]: Yeah I will create a better name apologies

File: Assets/Scripts/Map/TriggerSystem/Trigger Objects/Portal/PortalBehaviour.cs (hunk 3 above)
Comment (Dejin1) [reply Assets/Scripts/Map/TriggerSystem/Trigger Objects/Portal/PortalBehaviour.cs]: Thats hard to get out of habit doing the ==false

File: Assets/Scripts/Map/Weather/Weather.cs (hunk 5 above)
Comment (Dejin1) [reply Assets/Scripts/Map/Weather/Weather.cs]: do you need to if its one line?

File: Assets/Scripts/Player/JinPlayerMovement.cs (hunk 6 above)
Comment (Dejin1) [reply Assets/Scripts/Player/JinPlayerMovement.cs]: I thought you didn't need to if its only 1 line

File: Assets/Audio/AudioScripts/AudioFadeIn.cs [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+        {
+            audioSource.Stop();
```
[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Audio/AudioScripts/AudioFadeIn.cs]: Yes it is intended would take many conversations back and forth to explain in full but yes intended

File: Assets/Audio/AudioScripts/AudioFadeOut.cs [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+using System.Collections;
+
+public static class AudioFadeOut
+{
+
+    public static IEnumerator FadeOut(AudioSource audioSource, float FadeTime)
+    {
+        float startVolume = audioSource.volume;
+
+        while (audioSource != null && audioSource.volume > 0)
```
[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Audio/AudioScripts/AudioFadeOut.cs]: fixed in new commit

File: Assets/Audio/AudioScripts/AudioFadeIn.cs [hunk 11]
```diff
@@ -0,0 +1,25 @@
+﻿using UnityEngine;
+using System.Collections;
+
+public static class AudioFadeIn
```
[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Audio/AudioScripts/AudioFadeIn.cs]: I realized fade in was not used so it was deleted

File: Assets/Audio/AudioScripts/AudioFadeOut.cs [hunk 12]
```diff
@@ -0,0 +1,23 @@
+﻿using UnityEngine;
+using System.Collections;
+
+public static class AudioFadeOut
```
[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Audio/AudioScripts/AudioFadeOut.cs]: fixed in commit

File: Assets/Audio/AudioScripts/CallFade.cs [hunk 13]
```diff
@@ -0,0 +1,27 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class CallFade : MonoBehaviour
```
[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Audio/AudioScripts/CallFade.cs]: No its a class that is used to work with the FadeOut class

File: Assets/Scripts/Ai/ShrubbleTest.cs [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        anim.SetBool("Hide", true);
+    }
+    public void Idle()
+    {
+        anim.SetBool("Hide", false);
+    }
+    public void Run()
+    {
+        anim.SetBool("Run", true);
```
[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Scripts/Ai/ShrubbleTest.cs]: This is a demo script and will need to be removed later for a better AI its just for the demo I will put a ToDo on the script to delete it after new demo is not needed

File: Assets/Scripts/Ai/ShrubbleTest.cs [hunk 15]
```diff
@@ -0,0 +1,26 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class ShrubbleTest : MonoBehaviour
```
[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Scripts/Ai/ShrubbleTest.cs]: This class needs to be either replaced or deleted at the end of demo for a better Shrubble Ai its just in-place now for the demo to function. A to-do was added to define what i said to you.

File: Assets/Scripts/Animator/AnimatorResetBoolAtEnd.cs [hunk 16]
```diff
@@ -0,0 +1,18 @@
+﻿using UnityEngine;
+
+public class AnimatorResetBoolAtEnd : StateMachineBehaviour
```
[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Scripts/Animator/AnimatorResetBoolAtEnd.cs]: Fixed in commit

File: Assets/Scripts/Firing/Projectile.cs [hunk 17]
```diff
@@ -21,16 +21,21 @@ void Start()
 
     void OnTriggerEnter2D(Collider2D hitInfo)
     {
-        Destroy(gameObject);
+        if (hitInfo.tag == "Enemy" || hitInfo.tag == "Destructable")
```
[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Scripts/Firing/Projectile.cs]: I always miss these but i'm getting better fixed in new commit

File: Assets/Scripts/Firing/Projectile.cs [hunk 18]
```diff
@@ -21,16 +21,21 @@ void Start()
...
+        {
+            Enemy enemy = hitInfo.GetComponent<Enemy>();
```
[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Scripts/Firing/Projectile.cs]: I am not 100 percent sure how to do that?

File: Assets/Scripts/Firing/Projectile.cs [hunk 19]
```diff
@@ -21,16 +21,21 @@ void Start()
...
 
-        Enemy enemy = hitInfo.GetComponent<Enemy>();
+            if (enemy != null)
+            {
+                enemy.TakeDamage(damage);
+            }
 
-        if (enemy != null)
+            Instantiate(impactEffect, transform.position, transform.rotation);
```
[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Scripts/Firing/Projectile.cs]: The name ImpactEffect being initiated on a projectile should imply its for the projectile?

File: Assets/Scripts/Health/GainLife.cs [hunk 20]
```diff
@@ -20,9 +20,12 @@ void OnTriggerEnter2D (Collider2D col)
         {
             return;
         }
-
-        healthObject.GetComponent<Health>().ApplyHeal(addLife);  //Find the Health Script and apply damage = to the user set damage
-
+        if (healthObject.GetComponent<Health>().healthSlider.value < healthObject.GetComponent<Health>().healthSlider.maxValue)
```
[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Scripts/Health/GainLife.cs]: I'm not sure what to name them as health slider value is basically the value of the health slider and the healthObject is the healing object so i'm a bit confused how to make them more descriptive by renaming?

File: Assets/Scripts/Map/Portal/PortalAnimation.cs [hunk 21]
```diff
@@ -0,0 +1,81 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class PortalAnimation : MonoBehaviour
```
[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Scripts/Map/Portal/PortalAnimation.cs]: No it runs the logic to animate the Portal...so portal animation?

File: Assets/Scripts/Map/Portal/PortalAnimation.cs [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+    public static string strClosed = "Closed";
+    public static string strOpen = "Open";
+    public static string strIdle = "Idle";
+    public static string PortalClosed = "Portal_Closed";
+    public AudioSource opening;
+    public AudioSource idle;
+    public AudioSource closing;
+    private bool portalClosing = false;
+    private bool openTriggered;
+    public int h;
```
[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Scripts/Map/Portal/PortalAnimation.cs]: removed LOL

File: Assets/Scripts/Map/Portal/PortalAnimation.cs [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+    public float idleDelay = 0.00f;
+
+
+    AnimatorClipInfo[] animatorClipInfo;
+
+    public Animator animator;
+
+    public void Awake()
+    {
+        portalClosing = false;
```
[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Scripts/Map/Portal/PortalAnimation.cs]: fixed

File: Assets/Scripts/Map/TriggerSystem/Trigger Objects/Portal/PortalBehaviour.cs (hunk 1 above)
Comment (Dejin1) [reply Assets/Scripts/Map/TriggerSystem/Trigger Objects/Portal/PortalBehaviour.cs]: fixed

File: Assets/Scripts/Map/Weather/Weather.cs (hunk 4 above)
Comment (Dejin1) [reply Assets/Scripts/Map/Weather/Weather.cs]: removed not needed

File: Assets/Scripts/Map/Weather/Weather.cs (hunk 5 above)
Comment (Dejin1) [reply Assets/Scripts/Map/Weather/Weather.cs]: fixed

File: Assets/Scripts/Player/JinPlayerMovement.cs (hunk 6 above)
Comment (Dejin1) [reply Assets/Scripts/Player/JinPlayerMovement.cs]: fixed

**Comment by massieb** [comment]
@Dejin1 Why did you closed this merge request?

File: Assets/Audio/AudioScripts/CallFade.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (massieb) [reply Assets/Audio/AudioScripts/CallFade.cs]: But CallFade is a method name. That is not intended for a class. Unacceptable name for me. FadeCaller, FadeManager, FadePlayer. w/e but not CallFade.

File: Assets/Scripts/Ai/ShrubbleTest.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (massieb) [reply Assets/Scripts/Ai/ShrubbleTest.cs]: Todo is fine!

File: Assets/Scripts/Ai/ShrubbleTest.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (massieb) [reply Assets/Scripts/Ai/ShrubbleTest.cs]: Todo is fine

File: Assets/Scripts/Firing/Projectile.cs (hunk 18 above)
[the remark this answers is not in the document]
Comment (massieb) [reply Assets/Scripts/Firing/Projectile.cs]: Can you make a task for this. Then we will have a look later on together!

File: Assets/Scripts/Health/GainLife.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (massieb) [reply Assets/Scripts/Health/GainLife.cs]: float currentHealth = healthObject.GetComponent<Health>().healthSlider.value;
float maxHealth = healthObject.GetComponent<Health>().healthSlider.maxValue;
if(currentHealth < maxHealth) {
 ...
}

File: Assets/Scripts/Map/Portal/PortalAnimation.cs (hunk 21 above)
[the remark this answers is not in the document]
Comment (massieb) [reply Assets/Scripts/Map/Portal/PortalAnimation.cs]: Its not the animations. The animation is done with images. This is the logic to start/exectue those images to create an animation (am i right?). So you should name it after what it does.
PortalAnimationPlayer, PortalAnimationAnimator, PortalAnimationExecuter?

File: Assets/Scripts/Firing/Projectile.cs (hunk 18 above)
[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Scripts/Firing/Projectile.cs]: Yes can do will this be able to be fixed later if we make the change now as even though this change requested is a new feature the script now is better than it was before game wise even if it isn't optomised

File: Assets/Scripts/Health/GainLife.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Scripts/Health/GainLife.cs]: That makes sense apologies

[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Scripts/Health/GainLife.cs]: fixed in new demo 2

File: Assets/Scripts/Map/Portal/PortalAnimation.cs (hunk 21 above)
[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Scripts/Map/Portal/PortalAnimation.cs]: Hmm I see I in my mind it made sense but Im fine with that change fixed in demo 2

File: Assets/Audio/AudioScripts/CallFade.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (Dejin1) [reply Assets/Audio/AudioScripts/CallFade.cs]: Will do Fixed in Demo2

**Comment by Dejin1** [comment]
> @Dejin1 Why did you closed this merge request?

instead of merging from develop I kept the old changes and made it a new branch called demo 2 and fixed all the naming conventions you asked for there

File: Assets/Scripts/Firing/Projectile.cs (hunk 18 above)
[the remark this answers is not in the document]
Comment (massieb) [reply Assets/Scripts/Firing/Projectile.cs]: Yes. I saw you made the task so this is covered by that task

**Comment by massieb** [comment]
> > @Dejin1 Why did you closed this merge request?
> 
> instead of merging from develop I kept the old changes and made it a new branch called demo 2 and fixed all the naming conventions you asked for there

Unfortunately that is bot how it works. The reviews are documented here and i am not going to review the same changes in a second branch. 
I dont really understand why you started a new branch from this branch (as all the commits from here are also there). You could merge demo2 back into here to add your latest changes or merge develop into demo2 after this request is approved and merged to develop.

**Pull request closed without merging**